viewer/packages/cad-model/src/batching/MultiBufferBatchingManager.ts

```diff
@@ -0,0 +1,313 @@
+
+    const treeIndexInterleavedAttribute = this.getTreeIndexAttribute(sourceInstanceAttributes);
+
+    for (let i = 0; i < treeIndexInterleavedAttribute.count; i++) {
+      incrementOrInsertIndex(batchBuffer.mesh.userData.treeIndices, treeIndexInterleavedAttribute.getX(i));
+    }
+
+    if (bufferIsReallocated) {
+      this.reallocateBufferGeometry(batchBuffer);
+    } else {
+      this.updateInstanceAttributes(sourceInstanceAttributes, updateRange);
+    }
+
+    const instanceCount = sourceInstanceAttributes[0].attribute.count;
+    sectorInstanceData.instanceCount = instanceCount;
+
+    batchBuffer.mesh.visible = true;
+    batchBuffer.mesh.count += instanceCount;
+  }
+
+  private reallocateBufferGeometry({ buffer, mesh }: BatchBuffer) {
+    const defragmentedBufferGeometry = this.createDefragmentedBufferGeometry(mesh.geometry, buffer);
+    mesh.geometry.dispose();
+    mesh.geometry = defragmentedBufferGeometry;
+  }
+
+  private updateInstanceAttributes(
+    instanceAttributes: {
+      name: string;
+      attribute: InterleavedBufferAttribute;
+    }[],
+    updateRange: {
+      byteOffset: number;
+      byteCount: number;
+    }
+  ) {
+    instanceAttributes.forEach(namedAttribute => {
+      const attribute = namedAttribute.attribute;
+      this.extendUpdateRange(attribute, updateRange);
+      attribute.data.needsUpdate = true;
+    });
+  }
+
+  private extendUpdateRange(
+    attribute: THREE.InterleavedBufferAttribute,
+    updateRange: {
+      byteOffset: number;
+      byteCount: number;
+    }
+  ) {
+    const typeSize = (attribute.data.array as TypedArray).BYTES_PER_ELEMENT;
+    const newOffset = updateRange.byteOffset / typeSize;
+    const newCount = updateRange.byteCount / typeSize;
+
+    const { offset: oldOffset, count: oldCount } = attribute.data.updateRange;
+    if (oldCount === -1) {
+      attribute.data.updateRange = { offset: newOffset, count: newCount };
+      attribute.data.needsUpdate = true;
+      return;
+    }
+
+    const offset = Math.min(oldOffset, newOffset);
+    const count = Math.max(oldOffset + oldCount, newOffset + newCount) - offset;
+
+    attribute.data.updateRange = { offset, count };
+  }
+
+  private getBatchBufferToFill(instanceBatch: InstanceBatch) {
+    return instanceBatch.buffers.sort((a, b) => {
+      const aBufferSize = a.buffer.bufferView.length;
+      const bBufferSize = b.buffer.bufferView.length;
+      if (aBufferSize !== bBufferSize) {
+        return aBufferSize - bBufferSize;
+      }
+      return a.buffer.length - b.buffer.length;
+    })[0];
+  }
+
+  private getOrCreateInstanceBatch(parsedGeometry: Required<ParsedGeometry>) {
+    const instanceBatch = this._instanceBatches.get(parsedGeometry.instanceId);
+
+    if (instanceBatch !== undefined) {
+      return instanceBatch;
+    }
+
+    const batchBuffers: BatchBuffer[] = Array.from(Array(this.numberOfInstanceBatches)).map(() =>
```

Comment:
This will not work, due to js being wack `Array(n)` will give you an array of length n but will not have any values so you can't actually iterate over it...

Reply:
Here is an extra post if you really want to hate js even more: https://stackoverflow.com/questions/18947892/creating-range-in-javascript-strange-syntax